### PR TITLE
Tpwdcee 3783 post processing

### PR DESCRIPTION
**Pull request by brkicadis**

### CODE REVIEW

File: examples/payments/CreditCard/createUi.php [hunk 1]
```diff
@@ -30,7 +30,28 @@
 $transactionService = new TransactionService($config);
 
 $redirectUrl = getUrl('return.php?status=success');
-$amount = new Amount(70.00, 'EUR');
+
+$gatewayEnv = getenv('GATEWAY');
+
+if ('NOVA' == $gatewayEnv || 'API-TEST' == $gatewayEnv || 'API-WDCEE-TEST' == $gatewayEnv) {
```
Comment (tatsta) [inline examples/payments/CreditCard/createUi.php]: These 2 if's can be refactored

File: tests/_support/Page/CreditCardCreateUIAuthorization.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Shop System Plugins:
+ * - License can be found under:
+ * https://github.com/wirecard/paymentSDK-php/blob/master/LICENSE
+ */
+
+namespace Page;
+
+class CreditCardCreateUIAuthorization extends CreditCardCreateUiBase
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIAuthorization.php]: this class is extending CreditCardCreateUiBase class, but a lot of elements and methods are duplicated, we should use ones from the parent class

File: tests/_support/Page/CreditCardCreateUIPurchase.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Shop System Plugins:
+ * - License can be found under:
+ * https://github.com/wirecard/paymentSDK-php/blob/master/LICENSE
+ */
+
+namespace Page;
+
+class CreditCardCreateUIPurchase extends CreditCardCreateUiBase
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIPurchase.php]: same here for all the methods and comments

File: tests/_support/Page/CreditCardCreateUIAuthorization.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        'Save' => "//*[@class='btn btn-primary']",
+        'Credit Card payment form' => "//*[@id='payment-form']"
+    );
+
+    public $wirecard_frame = "wirecard-seamless-frame";
+
+    /**
+     * Method switchFrame
+     */
+    public function switchFrame()
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIAuthorization.php]: for example this method is duplicating the one from the parent class

File: tests/_support/Page/CreditCardCreateUIAuthorization.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        // Switch to Credit Card UI frame
+        $wirecard_frame = "wirecard-seamless-frame";
+        $I->executeJS('jQuery(".' . $this->wirecard_frame . '").attr("name", "' . $this->wirecard_frame . '")');
+        $I->switchToIFrame("$this->wirecard_frame");
+    }
+
+    /**
+     * Method prepareClick
+     */
+    public function prepareClick()
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIAuthorization.php]: same with this method

File: tests/_support/Page/CreditCardCreateUIAuthorization.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        parent::prepareClick();
+    }
+    /**
+     * Method Method prepareDataForField
+     * @param string $cardData
+     * @param null $type
+     */
+    public function fillCreditCardFields($cardData, $type = null)
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIAuthorization.php]: same here

File: tests/_support/Page/CreditCardCreateUIAuthorization.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Method prepareClick
+     */
+    public function prepareClick()
+    {
+        parent::prepareClick();
+    }
+    /**
+     * Method Method prepareDataForField
```
Comment (tatsta) [inline tests/_support/Page/CreditCardCreateUIAuthorization.php]: this method is called fillCreditCardFields , not prepareDataForField

File: tests/_support/Page/CreditCardPayBasedOnReserve.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    );
+
+    /**
+     * Method prepareDataForField
+     *
+     * @param string $fieldValue
+     * @param string $valueToKeepBetweenSteps
+     * @return string
+     */
+    public function prepareDataForField($fieldValue, $valueToKeepBetweenSteps)
```
Comment (tatsta) [inline tests/_support/Page/CreditCardPayBasedOnReserve.php]: this method is duplication the same on from CreditCardCancel.php, maybe we should create a parent a class or transfer this method to Base.php

File: tests/acceptance/Purchase3dsCheckPostProcessingPurchase.feature [hunk 9]
```diff
@@ -0,0 +1,39 @@
+Feature: checkCreditCard3DsFunctionalityPurchase
```
Comment (tatsta) [inline tests/acceptance/Purchase3dsCheckPostProcessingPurchase.feature]: Maybe we rename the file to be consistent to PurchaseCheckPostProcessingPurchase3ds.feature and Feature  to checkCreditCard3DsFunctionalityPostprocessingPurchase

File: tests/acceptance/PurchaseCheck3dsPostProcessingAuthorization.feature [hunk 10]
```diff
@@ -0,0 +1,81 @@
+Feature: checkCreditCard3DsFunctionalityAuthorization
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheck3dsPostProcessingAuthorization.feature]: also rename Feature to checkCreditCard3DsFunctionalityPostProcessingAuthorization

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorization.feature [hunk 11]
```diff
@@ -0,0 +1,61 @@
+Feature: checkCreditCardNon3DsFunctionalityAuthorization
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorization.feature]: checkCreditCardNon3DsFunctionalityPostProcessingAuthorization

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature [hunk 12]
```diff
@@ -0,0 +1,61 @@
+Feature: checkCreditCardNon3DsFunctionalityAuthorizationSg
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature]: same here

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature [hunk 13]
```diff
@@ -0,0 +1,29 @@
+Feature: checkCreditCardNon3DsFunctionalityPurchaseSg
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature]: and here

File: tests/acceptance/PurchaseNon3dsCheckPostProcessingPurchase.feature [hunk 14]
```diff
@@ -0,0 +1,29 @@
+Feature: checkCreditCardNon3DsFunctionalityPurchase
```
Comment (tatsta) [inline tests/acceptance/PurchaseNon3dsCheckPostProcessingPurchase.feature]: and here

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  As a user
+  I want to make a purchase with a Credit Card 3DS
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature]: non 3Ds

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+  As a user
+  I want to make a purchase with a Credit Card 3DS
```
Comment (tatsta) [inline tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature]: non 3DS

File: tests/acceptance/purchaseCheckHappyPath3ds.feature [hunk 17]
```diff
@@ -37,28 +37,3 @@ Feature: checkCrediCard3DSFunctionalityHappyPath
     And I click "Refund"
     Then I see text "Payment successfully cancelled."
     And I see text "Transaction ID"
-
-  @TEST-SG @SECURE-TEST-SG
```
Comment (tatsta) [inline tests/acceptance/purchaseCheckHappyPath3ds.feature]: maybe it's better to remove groups, and not the whole scenario, we might enable it when SG issues are solved

File: tests/acceptance/Purchase3dsCheckPostProcessingPurchase.feature (hunk 9 above)
Comment (brkicadis) [reply tests/acceptance/Purchase3dsCheckPostProcessingPurchase.feature]: Done

File: tests/acceptance/PurchaseCheck3dsPostProcessingAuthorization.feature (hunk 10 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheck3dsPostProcessingAuthorization.feature]: Done

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorization.feature (hunk 11 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorization.feature]: Done

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature (hunk 12 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature]: Done

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature (hunk 13 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature]: Done

File: tests/acceptance/PurchaseNon3dsCheckPostProcessingPurchase.feature (hunk 14 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseNon3dsCheckPostProcessingPurchase.feature]: Done

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature (hunk 15 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheckNon3dsPostProcessingAuthorizationSg.feature]: Done

File: tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature (hunk 16 above)
Comment (brkicadis) [reply tests/acceptance/PurchaseCheckNon3dsPostProcessingPurchaseSg.feature]: Done

File: tests/acceptance/purchaseCheckHappyPath3ds.feature (hunk 17 above)
Comment (brkicadis) [reply tests/acceptance/purchaseCheckHappyPath3ds.feature]: Done

File: tests/_support/Page/CreditCardCreateUIAuthorization.php (hunk 2 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIAuthorization.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardCreateUIPurchase.php (hunk 3 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIPurchase.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardCreateUIAuthorization.php (hunk 4 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIAuthorization.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardCreateUIAuthorization.php (hunk 5 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIAuthorization.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardCreateUIAuthorization.php (hunk 6 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIAuthorization.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardCreateUIAuthorization.php (hunk 7 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardCreateUIAuthorization.php]: It works without the rest of the duplicated stuff. Done

File: tests/_support/Page/CreditCardPayBasedOnReserve.php (hunk 8 above)
Comment (brkicadis) [reply tests/_support/Page/CreditCardPayBasedOnReserve.php]: It works without the rest of the duplicated stuff. Done

Comment (brkicadis) [reply tests/_support/Page/CreditCardPayBasedOnReserve.php]: Method moved to Base.php. Done

File: examples/payments/CreditCard/createUi.php [hunk 18]
```diff
@@ -30,7 +30,26 @@
...
 
 $redirectUrl = getUrl('return.php?status=success');
-$amount = new Amount(70.00, 'EUR');
+
+$gatewayEnv = getenv('GATEWAY');
+
+$currency='EUR';
+$startAmount = 25;
+
+if (strpos($gatewayEnv, 'SG')>0) {
```
Comment (pretzelhands) [inline examples/payments/CreditCard/createUi.php]: `strpos` may potentially return 0 if the substring was found at the very beginning. If no string was found whatsoever it returns `false`.

File: examples/payments/CreditCard/createUi.php [hunk 19]
```diff
@@ -30,7 +30,26 @@
...
+if (70 == intval($_GET['amount'])) {
+    $startAmount = 70;
+}
+
+$amount = new Amount($startAmount, $currency);
+
+if ($_SERVER['REQUEST_METHOD'] == 'POST') {
+    $postedAmount = $_POST['amount'];
+    $currency = $_POST['currency'];
+    $amount = new Amount((int)$postedAmount, $currency);
```
Comment (pretzelhands) [inline examples/payments/CreditCard/createUi.php]: For consideration: Should we support amounts like `29.50`? Currently that value would be cast to just `29`. I'm not sure if this is relevant for testing purposes.

File: examples/payments/CreditCard/reserve.php [hunk 20]
```diff
@@ -23,7 +23,7 @@
 // ### Transaction related objects
 
 // Create a amount object as amount which has to be paid by the consumer.
-$amount = new Amount(70.00, 'EUR');
+$amount = new Amount((int)$_POST['amount'], $_POST['currency']);
```
Comment (pretzelhands) [inline examples/payments/CreditCard/reserve.php]: Same as above, do we never require float amounts for testing?

File: tests/_support/Page/Base.php [hunk 21]
```diff
@@ -63,21 +63,27 @@ public function prepareClick()
...
-        return $fieldValue;
+        if (strpos($fieldValue, "Noted") !== false) {
+            return $valueToKeepBetweenSteps;
+        } else {
+            return $fieldValue;
+        }
     }
 
     /**
      * Method Method prepareDataForField
```
Comment (pretzelhands) [inline tests/_support/Page/Base.php]: "Method method" seems a bit redundant. :)

File: tests/_support/Page/Base.php [hunk 22]
```diff
@@ -63,21 +63,27 @@ public function prepareClick()
...
      * @param string $cardData
      */
-    public function fillCreditCardFields($cardData){
+    public function fillCreditCardFields($cardData)
+    {
         ;
```
Comment (pretzelhands) [inline tests/_support/Page/Base.php]: Is this function used anywhere? It doesn't seem incredibly useful from a quick glance.

**Pull request closed without merging**